[clang][DependencyScanning] Remove dependency on clangDriver from clangDependencyScanning

naveen-seth opened the pull request:
This is the final patch in a series that removes the dependency of `clangDependencyScanning` on `clangDriver`, splitting the work from #169964 into smaller changes (see comment linked below).

This patch updates the remaining parts of the scanning interface in `DependencyScanningWorker` to only operate on `-cc1` / driver job command lines.
This follows #171238, which applied this change to the by-name scanning API.

This is part of a broader effort to support driver-managed builds for compilations using C++ named modules and/or Clang modules. It is required for linking the dependency scanning tooling against the driver without introducing cyclic dependencies, which would otherwise cause build failures when dynamic linking is enabled.

naveen-seth committed:
[clang][DependencyScanning] Remove dependency on clangDriver from clangDependencyScanning

This is the final patch in a series that removes the dependency of
clangDependencyScanning on clangDriver, splitting the work
from #169964 into smaller changes (see comment linked below).

This patch updates the remaining parts of the scanning interface in
DependencyScanningWorker to only operate on -cc1 / driver job command
lines.
This follows #171238, which applied this change to the by-name scanning
API.

This is part of a broader effort to support driver-managed builds for
compilations using C++ named modules and/or Clang modules. It is
required for linking the dependency scanning tooling against the
driver without introducing cyclic dependencies, which would otherwise
cause build failures when dynamic linking is enabled.

jansvoboda11 commented on clang/include/clang/DependencyScanning/DependencyScanningWorker.h:
```diff
@@ -93,29 +93,39 @@ class DependencyScanningWorker {
-  /// TUBuffer is not nullopt, it is used as TU input for the dependency
-  /// scanning. Otherwise, the input should be included as part of the
-  /// command-line.
+  /// Run the dependency scanning tool for the given driver job command-line,
```
Thanks for updating the documentation, this is great! Suggestion is to avoid confusion between the worker/tool and driver/frontend command lines.

```suggestion
  /// Run the dependency scanning worker for the given frontend command-line,
```

jansvoboda11 commented on clang/include/clang/DependencyScanning/DependencyScanningWorker.h:
```diff
@@ -93,29 +93,39 @@ class DependencyScanningWorker {
+  bool computeDependencies(
+      StringRef WorkingDirectory, ArrayRef<ArrayRef<std::string>> CommandLines,
+      DependencyConsumer &DepConsumer, DependencyActionController &Controller,
+      DiagnosticsEngine &Diags,
```
Can you explain why we're taking the entire engine instead of just the consumer like above?

naveen-seth committed:
Apply suggestions from code review

Co-authored-by: Jan Svoboda <jan@svoboda.ai>

naveen-seth committed:
Merge branch 'main' into remove-driver-dep-from-dependency-scanning

jansvoboda11 commented on clang/lib/Tooling/DependencyScanningTool.cpp:
```diff
@@ -74,13 +79,131 @@ class MakeDependencyPrinterConsumer : public DependencyConsumer {
+                 IntrusiveRefCntPtr<llvm::vfs::FileSystem> FS,
+                 llvm::BumpPtrAllocator &Alloc) {
+  auto Argv = llvm::map_to_vector<256>(
+      CommandLine, [](const auto &Arg) { return Arg.c_str(); });
```
I'd avoid this refactor here and move it into a separate commit. (TBH my personal preference would be to keep the original loop version, since it's simpler.)

jansvoboda11 reviewed (commented):
This is great, splitting out a couple of smaller PRs makes this a breeze to review! I only have minor feedback. Let's see what @qiongsiwu has to say too.

naveen-seth committed:
Merge remote-tracking branch 'upstream/main' into remove-driver-dep-from-dependency-scanning

naveen-seth committed:
Revert buildCompilation changes

naveen-seth committed:
Move away from map_to_vector, prefer SmallVector

Now that we are using ArrayRef consistently in the computeDependencies
function signatures, we can also use SmallVector over std::vector as
per
https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h

naveen-seth committed:
Inline scanDependencies

naveen-seth committed:
auto use

naveen-seth committed:
Align computeDependencies signatures, use new DiagnosticEngine for each frontend invocation.

naveen-seth committed:
Formatting

jansvoboda11 reviewed (approved):
LGTM with minor suggestions/questions. Great job, thank you!

naveen-seth committed:
Move TUBuffer handling to getTranslationUnitDependencies

naveen-seth committed:
Undo renaming to buildFrontendCommandLine

naveen-seth committed:
Merge remote-tracking branch 'upstream/main' into remove-driver-dep-from-dependency-scanning

naveen-seth commented:
I’m neutral to slightly positive on this for now, but in the long term I think this is a good change.

In https://github.com/llvm/llvm-project/pull/152770#issuecomment-3344634967 and the reply https://github.com/llvm/llvm-project/pull/152770#issuecomment-3353740372 there is some discussion about unifying more of the dependency-scanning use-cases.

I feel like the driver could be a good place to do this.
Maybe we could even have a single function to drive the concurrent dependency scan and have it be reused by `clang-scan-deps`, `clangd`, `-fmodules-driver`, ….

Doing this change might be unnecessary if it turns out that unifying dependency scanning makes more sense somewhere else, but I can imagine it will take some time to reach a good decision, so I’m not too worried about it being premature either.

naveen-seth committed:
Merge branch 'main' into remove-driver-dep-from-dependency-scanning

naveen-seth commented:
Ping @qiongsiwu

qiongsiwu commented on clang/lib/Tooling/DependencyScanningTool.cpp:
```diff
@@ -74,14 +79,124 @@ class MakeDependencyPrinterConsumer : public DependencyConsumer {
+      DiagPrinterWithOS.DiagnosticsOS.str(), llvm::inconvertibleErrorCode());
+}
+
+static bool computeDependencies(
```
I think we want to make this a public interface available through the header `clang/include/clang/DependencyScanning/DependencyScanningWorker.h`. The reason is that we will need to call it for the CIndex APIs only available downstream for Swift. See https://github.com/swiftlang/llvm-project/blob/657ddd43fca3a2924d067f3930f370bd7e3a842b/clang/tools/libclang/CDependencies.cpp#L370. The `Compilation` argument is not really an instance of `driver::Compilation` but just an array of strings, which can be a driver command instead of a CC1 command.

qiongsiwu reviewed (commented):
Thanks for the ping! Sorry I missed the notifications on this review. 

The changes overall look great! I have one suggestion that facilitates merging this PR to the downstream Swift's llvm fork.

naveen-seth committed:
Expose computeDependencies in DependencyScanningTool.h

qiongsiwu reviewed (approved):
LGTM! Thanks!

naveen-seth commented:
> LGTM! Thanks!

Thanks for reviewing!

naveen-seth commented:
> I think this change might be responsible for the failures here: https://lab.llvm.org/buildbot/#/builders/25/builds/14594
> 
> I think they are hitting a UBSan check:
> 
> …

I believe this was fixed by #174515. The corresponding Buildbot build (https://lab.llvm.org/buildbot/#/builders/25/builds/14595) is not hitting the UBSan checks anymore.
